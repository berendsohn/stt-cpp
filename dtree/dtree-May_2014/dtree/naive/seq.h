// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_SEQ_H_
#define DTREE_NAIVE_SEQ_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndSeq : public Base {
 public:
  EndSeq() : solid_() {}
  EndSeq* solid(int i) const { return solid_[i]; }
  void set_solid(int i, EndSeq* solid_i) { solid_[i] = solid_i; }
  template<typename Forwarder>
  void CopyFrom(const EndSeq& node, Forwarder forward);
 private:
  EndSeq* solid_[2];
  DISALLOW_COPY_AND_ASSIGN(EndSeq);
};

template<typename Group, typename Base>
class WithValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToSeq(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromSeq(Node* node, typename Group::Type delta);

 protected:
  WithValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  template<typename Node>
  static typename Semigroup::Type AggrSeq(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostSeq(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeftmostSeq(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRightmostSeq(Node* node, const Predicate& predicate);
 protected:
  WithAggr() {}
};

template<typename Base>
class WithReverseBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverseBy() {}
};

template<typename Base>
class WithReverse : public Base {
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverse() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_SEQ_H_
