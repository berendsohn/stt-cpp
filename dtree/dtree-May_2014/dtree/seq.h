// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_SEQ_H_
#define DTREE_SEQ_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/seq.h"
#else
#include "dtree/selfadjust/seq.h"
#endif
#endif  // DTREE_SEQ_H_
