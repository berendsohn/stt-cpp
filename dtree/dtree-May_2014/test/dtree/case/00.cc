#define DTREE_CONFIG EndSeq<WithReverse<WithAggr<dtree::DpMax<double>, dtree::DwGreater_<double>, WithValue<dtree::DpAdd<double>, WithValue<dtree::Nop<int>, Begin > > > > >
#include "test/dtree/seq.cc"
