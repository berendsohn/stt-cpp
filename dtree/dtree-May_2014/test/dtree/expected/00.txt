<<<
  2: v2 vL3R-12F0
  4: v3 vL7R4F1

  3: v3 vL-2R1F0
  8: v7 vL-1R2F0
  5: v5 vL2R0F0

  6: v6 vL-3R-9F0

  9: v5 vL-6R-4F1
  1: v0 vL-45R2F1
  11: v4 vL-1R2F1
  15: v5 vL-10R-7F1
  12: v4 vL2R-23F1
  13: v4 vL-3R1F0

  14: v4 vL-10R-3F1
  10: v1 vL-9R-2F1
  7: v0 vL-1R2F1
>>>
<<<
    /8: v7 vL1R1F0 aL3R0 f1
      \5: v5 vL3R-2F0 aL0R0 f1
  @3: v3 vL-2R1F0 aL0R0 f1

    /2: v2 vL-16R-4F1 aL0R0 f0
  @4: v3 vL7R4F1 aL0R0 f0

  @6: v6 vL-3R-9F0 aL0R0 f0

  @13: v4 vL-3R1F0 aL2R1 f1
    \12: v4 vL5R-24F1 aL25R0 f0
        /15: v5 vL-9R35F0 aL9R9 f0
          \11: v4 vL9R9F0 aL0R0 f1
      \1: v0 vL25R-47F0 aL0R44 f0
        \9: v5 vL-6R39F0 aL0R0 f0

  @14: v4 vL-10R-3F1 aL5R12 f0
      /10: v1 vL-4R-8F0 aL0R0 f1
    \7: v0 vL5R9F0 aL0R0 f0
>>>
